components/core/tests/test-StreamingReader.cpp

```diff
@@ -0,0 +1,161 @@
+}
+
+TEST_CASE("streaming_reader_with_offset_and_seek", "[StreamingReader]") {
+    constexpr size_t cOffset{319};
+    clp::FileReader ref_reader;
+    ref_reader.open(get_ref_file_abs_path().string());
+    ref_reader.seek_from_begin(cOffset);
+    std::vector<char> ref_data;
+    read_into_memory_buffer(ref_reader, ref_data);
+    ref_reader.close();
+
+    REQUIRE((clp::ErrorCode_Success == clp::StreamingReader::init()));
+    std::vector<char> streamed_data;
+
+    // Read by opening with the offset.
+    {
+        clp::StreamingReader reader_using_offset(cTestUrl, cOffset);
+        read_into_memory_buffer(reader_using_offset, streamed_data);
+        auto const ret_code_using_offset{reader_using_offset.get_curl_return_code()};
+        REQUIRE(ret_code_using_offset.has_value());
+        // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+        REQUIRE((CURLE_OK == ret_code_using_offset.value()));
+        REQUIRE((streamed_data == ref_data));
+        streamed_data.clear();
+    }
+
+    // Read by seeking to the offset.
+    {
+        clp::StreamingReader reader_using_seek(cTestUrl);
+        reader_using_seek.seek_from_begin(cOffset);
+        read_into_memory_buffer(reader_using_seek, streamed_data);
+        auto const ret_code_using_seek{reader_using_seek.get_curl_return_code()};
+        REQUIRE(ret_code_using_seek.has_value());
+        // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+        REQUIRE((CURLE_OK == ret_code_using_seek.value()));
+        REQUIRE((streamed_data == ref_data));
+        streamed_data.clear();
+    }
+
+    clp::StreamingReader::deinit();
+}
+
+TEST_CASE("streaming_reader_destruct", "[StreamingReader]") {
+    REQUIRE((clp::ErrorCode_Success == clp::StreamingReader::init()));
+
+    // NOLINTBEGIN(cppcoreguidelines-avoid-magic-numbers,readability-magic-numbers)
+    bool peacefully_destructed{false};
+    auto test_abort = [&]() -> void {
+        // We sleep for a while to fill out all the buffers, and we delete the reader. The
+        // destructor should be called to abort the underlying transfer session. We should ensure
+        // destructor is successfully executed without deadlock or exceptions in this case.
+        try {
+            auto reader{std::make_unique<clp::StreamingReader>(
+                    cTestUrl,
+                    0,
+                    true,
+                    clp::StreamingReader::cDefaultOverallTimeout,
+                    clp::StreamingReader::cDefaultConnectionTimeout,
+                    3,
+                    512
+            )};
+            std::this_thread::sleep_for(std::chrono::seconds{1});
+            REQUIRE(reader->is_curl_transfer_in_progress());
+            reader.reset(nullptr);
+        } catch (clp::StreamingReader::OperationFailed const& ex) {
+            return;
+        }
+        peacefully_destructed = true;
+    };
+
+    REQUIRE(run_with_timeout(std::chrono::milliseconds{1500}, test_abort));
```

Comment:
Sorry, forgot to mention this in the previous review:

Do we absolutely need this async + timeout logic?

* `std::async` can run things in background thread which complicates the curl deinit logic in older versions of curl.
* Using performance sensitive code (like a timeout) can cause issues in GH's macOS runners since they have very inconsistent performance.
* If the only goal of using `run_with_timeout` is to avoid having the unit tests freeze, I would say let's just let the unit tests freeze. We will still detect that they freeze and can debug why.

Reply:
Have async removed